implement a modular structure for metadata backends

zoltan opened the pull request:
currently the foundationdb backend if built as if it were a real shared library. however, this can only be used from within hyperwarp, and is not reusable (being tightly coupled to metadata itself), while we want our metadata library to be actually not coupled to any particular backend.

this change implements this idea.

zoltan committed:
add modular metadata support

Signed-off-by: Zoltan Arnold Nagy <zoltan.arnold.nagy@gmail.com>

zoltan committed:
get rid of the gmodule dependency - we don't use its features right now anyway

Signed-off-by: Zoltan Arnold Nagy <zoltan.arnold.nagy@gmail.com>

zoltan committed:
make the foundationdb backend a real module instead of a shared library

Signed-off-by: Zoltan Arnold Nagy <zoltan.arnold.nagy@gmail.com>

zoltan commented:
rebased on latest master

nielspardon commented on metadata-foundationdb/metadata-foundationdb.h:
```diff
@@ -28,7 +25,7 @@ void metadata_persist(FDBDatabase *database, const Metadata *meta_data);
  * \retval NULL if the MetaData could not be found or if there was an error unpacking
  */
-Metadata *metadata_get(FDBDatabase *database);
+Metadata *metadata_get();
```
you renamed that to `load_metadata()` above

zoltan replied:
good catch, thanks

nielspardon commented on samples/metadata-sample.c:
```diff
@@ -70,29 +55,28 @@ int main()
     print_metadata(metadata, 0);
 
-    metadata_persist(database, metadata);
+    //metadata_persist(database, metadata);
```
can you add the new way of persisting metadata here and not just comment out the old way?

nielspardon commented on samples/metadata-sample.c:
```diff
@@ -70,29 +55,28 @@ int main()
     free(allocator);
 
-    metadata = metadata_get(database);
+    //metadata = metadata_get(database);
```
can you add the new way of loading metadata here and not just comment out the old way?

nielspardon commented on metadata-foundationdb/metadata-foundationdb.c:
```diff
@@ -196,4 +198,35 @@ void virtual_disk_persist(FDBDatabase *database, VirtualDisk *virtual_disk)
+    chk(fdb_create_database(NULL, &foundationdb_database));
+}
+
+MetadataBackend foundationdb_backend = {
+    .name = "foundationdb",
+    .initialize = foundationdb_backend_initialize,
+    .load = load_metadata,
+    .persist = metadata_persist
+};
```
I think we also need a `foundationdb_backend_finalize` in order to destroy the database and stop the network thread.

nielspardon commented on samples/metadata-sample.c:
```diff
@@ -70,29 +55,28 @@ int main()
 
-    print_metadata(metadata, 0);
+    //print_metadata(metadata, 0);
 
-    metadata__free_unpacked(metadata, NULL);
+    //metadata__free_unpacked(metadata, NULL);
```
I think this should be uncommented again.

nielspardon commented on samples/metadata-sample.c:
```diff
@@ -70,29 +55,28 @@ int main()
-    fdb_database_destroy(database);
+    //fdb_database_destroy(database);
 
-    chk(fdb_stop_network());
-    pthread_join(net_thread, NULL);
+    //chk(fdb_stop_network());
+    //pthread_join(net_thread, NULL);
```
This should be implemented by the `foundationdb_backend_finalize` mentioned above.

nielspardon commented on metadata-foundationdb/metadata-foundationdb.c:
```diff
@@ -3,10 +3,15 @@
 
 uint64_t metadata_key = 0ULL;
 
+FDBDatabase *foundationdb_database = NULL;
```
should we prefix this with `g_`?

zoltan replied:
I originally took that prefix from some of the glib example code.

Looking up visibility rules in C, it seems like the best option is to declare it static - this should restrict visibility to the compilation unit, eg. metadata-foundationdb.c itself.

nielspardon commented on metadata/CMakeLists.txt:
```diff
@@ -17,12 +17,13 @@
                           "${PROJECT_BINARY_DIR}"
                           "${PROJECT_SOURCE_DIR}"
                           "${CMAKE_CURRENT_SOURCE_DIR}"
+                          "${GLIB2_MODULE_INCLUDE_DIRS}"
```
seems you are not using glib2 anymore so I guess this can be removed

zoltan replied:
indeed, as discussed, I've switched to pure dlopen() as for right now it works for us equally both.

nielspardon commented on metadata-foundationdb/metadata-foundationdb.c:
```diff
@@ -108,9 +110,9 @@ void metadata_persist(FDBDatabase *database, const Metadata *metadata)
  * \retval NULL if the MetaData could not be found or if there was an error unpacking
  */
-Metadata *metadata_get(FDBDatabase *database)
+Metadata *load_metadata()
```
maybe this should be called `metadata_load` to follow the same pattern of `<object>_<verb>` as with `metadata_persist` above

zoltan replied:
since the module is explicitly tied to metadata, we could get rid of the first part, and just leave the verb.

nielspardon commented on metadata-foundationdb/CMakeLists.txt:
```diff
@@ -14,17 +14,17 @@
-set_target_properties(hyperwarp-metadata-fdb PROPERTIES PUBLIC_HEADER metadata-foundationdb.h)
-install(TARGETS hyperwarp-metadata-fdb LIBRARY PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/hyperwarp)
+set_target_properties(metadata-foundationdb PROPERTIES PREFIX "")
+install(TARGETS metadata-foundationdb LIBRARY DESTINATION ${PROJECT_BINARY_DIR})
```
I think we need to revisit whether `${PROJECT_BINARY_DIR}` is the right value here for ´make install´.

zoltan replied:
I'm open to suggestions. right now I never do make install.

nielspardon replied:
what about `${CMAKE_INSTALL_LIBDIR}/hyperwarp` which would result in something like `/usr/local/lib/hyperwarp`?

zoltan replied:
sounds good

nielspardon reviewed (changes requested):
a couple of comments / changes

zoltan committed:
fix discrepancy between header definition and actual function definition

Signed-off-by: Zoltan Arnold Nagy <zoltan.arnold.nagy@gmail.com>

zoltan committed:
declare global variables as static to make their visibility explicit

Signed-off-by: Zoltan Arnold Nagy <zoltan.arnold.nagy@gmail.com>

zoltan committed:
remove old and unused include

Signed-off-by: Zoltan Arnold Nagy <zoltan.arnold.nagy@gmail.com>

zoltan committed:
remove the metadata-foundationdb.h header

all metadata backends are modules only used by the metadata library via
the function pointers in the module descriptor. this means they will never
explicitly be included anywhere, so no need for a header.

Signed-off-by: Zoltan Arnold Nagy <zoltan.arnold.nagy@gmail.com>

zoltan committed:
make functions really local to the foundationdb module

using static makes them only available within the compilation unit,
eg. the foundationdb  module itself; so no name clash can happen.

Signed-off-by: Zoltan Arnold Nagy <zoltan.arnold.nagy@gmail.com>

zoltan committed:
implement persist and load properly in the foundationdb metadata backend

Signed-off-by: Zoltan Arnold Nagy <zoltan.arnold.nagy@gmail.com>

zoltan commented:
the new changes should resolve all your requested changes. I've tested it with a local FDB server and it worked.

nielspardon commented on samples/metadata-sample.c:
```diff
@@ -5,35 +5,23 @@
+    metadata_backend_initialize();
 
-    Metadata *metadata = new_metadata();
     Allocator *allocator = create_allocator(VIRTUAL_DISK__ERASURE_CODE_PROFILE__EC_4_2P);
```
I would move the `Metadata *metadata = new_metadata();` back down after the `metadata_backend_initialize();` since we only need to start allocating memory if the backend initialization was successful and we have loaded the metadata backend successfully. Otherwise we are leaking some memory if we abort while setting up the backend.

nielspardon reviewed (changes requested):
I think this is looking way better now. There's one small issue with possible memory leakage that should be fixed.

zoltan committed:
only allocate the metadata structure if the backend is available

Signed-off-by: Zoltan Arnold Nagy <zoltan.arnold.nagy@gmail.com>

zoltan committed:
make the module installation directory more sane

Signed-off-by: Zoltan Arnold Nagy <zoltan.arnold.nagy@gmail.com>

nielspardon commented on samples/metadata-sample.c:
```diff
@@ -5,35 +5,24 @@
 
-    pthread_t net_thread;
-    assert(0 == pthread_create(&net_thread, NULL, run_net, NULL));
+    Metadata *metadata = new_metadata();
```
that's already better however I thought about calling it after `metadata_backend_initialize();` two lines further below

zoltan replied:
you're right. I've now improved the error handling. Please take a look.

zoltan committed:
improve error handling in the metadata sample code

Signed-off-by: Zoltan Arnold Nagy <zoltan.arnold.nagy@gmail.com>

nielspardon reviewed (approved):
LGTM if you now could squash the commits into one before merging :-)
